conjure-java-server-verifier/src/test/java/com/palantir/verification/server/AutoDeserializeTest.java

```diff
@@ -20,87 +20,81 @@
-    public String jsonString;
-
-    @Parameterized.Parameters(name = "{0}({3}) -> should succeed {2} with client {4}")
-    public static Collection<Object[]> data() {
-        List<Object[]> objects = new ArrayList<>();
-        Cases.TEST_CASES.getAutoDeserialize().forEach((endpointName, positiveAndNegativeTestCases) -> {
-            int positiveSize = positiveAndNegativeTestCases.getPositive().size();
-            int negativeSize = positiveAndNegativeTestCases.getNegative().size();
-
-            IntStream.range(0, positiveSize)
-                    .forEach(i -> objects.add(new Object[] {
-                        endpointName,
-                        i,
-                        true,
-                        positiveAndNegativeTestCases.getPositive().get(i),
-                    }));
-
-            IntStream.range(0, negativeSize)
-                    .forEach(i -> objects.add(new Object[] {
-                        endpointName,
-                        positiveSize + i,
-                        false,
-                        positiveAndNegativeTestCases.getNegative().get(i)
-                    }));
-        });
-        return objects;
+    public static final class Parameters implements ArgumentsProvider {
+
+        @Override
+        public Stream<? extends Arguments> provideArguments(ExtensionContext _context) {
+            return Cases.TEST_CASES.getAutoDeserialize().entrySet().stream().flatMap(testCase -> {
+                EndpointName endpointName = testCase.getKey();
+                PositiveAndNegativeTestCases positiveAndNegativeTestCases = testCase.getValue();
+
+                int positiveSize = positiveAndNegativeTestCases.getPositive().size();
+                int negativeSize = positiveAndNegativeTestCases.getNegative().size();
+
+                return Stream.concat(
+                        IntStream.range(0, positiveSize)
+                                .mapToObj(i -> Arguments.of(
+                                        endpointName,
+                                        i,
+                                        true,
+                                        positiveAndNegativeTestCases
+                                                .getPositive()
+                                                .get(i))),
+                        IntStream.range(0, negativeSize)
+                                .mapToObj(i -> Arguments.of(
+                                        endpointName,
+                                        positiveSize + i,
+                                        false,
+                                        positiveAndNegativeTestCases
+                                                .getNegative()
+                                                .get(i))));
+            });
+        }
     }
 
-    @Test
-    public void runTestCase() {
+    @ParameterizedTest(name = "{0}({3}) -> should succeed {2} with client {4}")
+    @ArgumentsSource(Parameters.class)
```

Comment:
TIL: I hadn't used class-based argument sources, but I like it a lot better than specifying a method name as a string. Thanks!

Reply:
Oh ya, I've always used ArgumentsProvider instead of the string based method. Typesafety is so much nicer, and the string based methods _only_ make sense if you're using a language like Groovy which support table based spec testing.